### ISSUE TITLE
[Regression] False positive in nullable analysis of local functions

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#71419]
**Version Used**: 
```
commit 182e829faf57b84150a2b28036f20918f368fc9d
Merge: 944f84f460e 837cfcc0ad8
Author: Jason Malinowski <jason.malinowski@microsoft.com>
Date:   Thu Dec 21 21:45:44 2023 -0800

    Merge pull request #71395 from jasonmalinowski/fix-uri-creation-errors-for-generated-documents

    Call GetURI when getting a URI for a document
```

**Steps to Reproduce**:

Compile the following code:
```cs
#nullable enable
class C
{
  string? F;

  void M2()
  {
    local1();

    void local2()
    {
      F.ToString(); // CS8602: Dereference of a possibly null reference.
    }

    void local1()
    {
      F = "";
      local2();
    }
  }
}
```

[sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4GIYFcA2OLA4CmABETAcQLABQAAgEwkDCtA3rSSXQIwAMAfhIAxANy1O3ACwkAsgwAUASkkcaXLjgD2AYwg4ey8TUlc6M7XpyKV6jWo0bhAOgAqWgMoAXAE6wA5kYkAPTBLB4AHABsfAwgJAAiRD5EAGbJ5DqkWqkkECQADloAzsVQhACeJNh4JCnpKTBZzqYkAL4SdmYWuvqGto4OjiIkALwkAEQTxsMklvo2MxoddittQA)

**Diagnostic Id**: CS8602

**Expected Behavior**:
No warnings.
`local2` is only called from `local1` which takes care of checking/initializing `F` to a non-nullable value before invoking it

**Actual Behavior**:
warning `CS8602: Dereference of a possibly null reference` is reported for the dereference of `F` in `local2`

**Notes**:
The code builds without warnings in the latest released VS (`VisualStudio.17.Release/17.8.3+34330.188`) and the latest preview VS (`VisualStudio.17.Preview/17.9.0-pre.2.0+34407.89`)

I believe this is a regression introduced in https://github.com/dotnet/roslyn/pull/71148 by making a default state of a local function with no _known_ call sites reachable. Consequently, instead of being discarded, this default state is merged with any reachable call site state that might be discovered later while inspecting other local functions.

**Comment by AlekseyTs** [comment]
@jjonescz Is this fixed already?

**Comment by jjonescz** [comment]
> Is this fixed already?

No, this is really a regression introduced by my fix https://github.com/dotnet/roslyn/pull/71148.
